### PR TITLE
Support for Zvk, Vector Cryptography Extensions

### PR DESCRIPTION
**Pull request by egouriou-rivos**
Support for Zvk, Vector Cryptography Extensions
   
Add encodings for all instructions in the **unratified** Zvk extensions:
 - Zvbb, Vector Bit-manipulation instructions used in Cryptography,
 - Zvbc, Vector Carryless Multiplication
 - Zvkg, Vector GCM/GMAC instruction for Cryptography,
 - Zvkned, NIST Suite: Vector AES Encryption & Decryption (Single
   Round),
 - Zvknha, Zvknhb, NIST Suite: Vector SHA-2,
 - Zvksed, ShangMi Suite: SM4 Block Cipher Instructions
 - Zvkssh, ShangMi Suite: SM3 Hash Function Instructions

Add two "shorthand" extensions:
 - Zvkn: NIST Suite, imports Zvbb, Zvbc, Zvkned, and Zvknh
 - Zvks: ShangMi Suite, imports Zvbb, Zvbc, Zvksed, and Zvksh

Note that we do *not* list the Zvkt extension,  as it does not contain
instructions itself, only implies that instructions in various parts of the ISA
respect the data-independent execution latency (DIEL) constraint.

Three new fields are listed in constants.py:
 - 'zimm5', used to encode round constants (Zvkns, Zvksed, Zvksh),
   and 5-bit shift constant (vwsll.vi in Zvbb)
 - 'zimm6hi, zimm6lo', used to encode the 6 bits rotate amount
   in vror.vi.

The Zvk instructions – with the exception of Zvbb, Zvbc – reside in the
P opcode space. Some encodings conflict with proposed instructions
in the P extension (packed SIMD). Zvk and P are exclusive of each
other, no implementation will implement both. Conflicting P instructions
are marked as pseudo of the Zvk instructions.

The encodings match the current documentation of the specification
at <https://github.com/riscv/riscv-crypto/tree/master/doc/vector>,
at Version v0.4.6, 22 March, 2023.

Co-authored-by: Eric Gouriou <ego@rivosinc.com>
Co-authored-by: Stanislaw Kardach <kda@semihalf.com>
Co-authored-by: Kornel Duleba <mindal@semihalf.com>
Co-authored-by: Raghav Gupta <rgupta@rivosinc.com>
Signed-off-by: Eric Gouriou <ego@rivosinc.com>

### CODE REVIEW

**Comment by egouriou-rivos** [comment]
@aswaterman Andrew, this is my first PR against this repo, I am not sure what the exact process is to request review/approval. Don't hesitate to let me know what is best for you.

**Comment by egouriou-rivos** [comment]
@kdockser FYI

**Comment by aswaterman** [comment]
@egouriou-rivos it's actually @neelgala who should review your PR, not me.

File: unratified/rv_zvknh
Comment (neelgala) [inline unratified/rv_zvknh:1]: The spec diffrentiates between zvknha and zvknhb - indicating there is not extension named zvknh. In that case these instructions must be split into multiple files - rv_zvknha and rv_zvknhb

**Review by neelgala — CHANGES_REQUESTED** [review]

File: unratified/rv_zvknh
Comment (egouriou-rivos) [reply unratified/rv_zvknh]: Done.

**Comment by egouriou-rivos** [comment]
Neel, thanks for the suggestion.

I updated the commit with the following changes:
(1) I split Zvknha/Zvknhb into separate files, per Neel's request.
Per the README guidelines, I used Zvknha to define the instructions, and have them imported into Zvknhb. I think import is the right mechanism (vs pseudo_op), but please let me know if you feel differently.
 
(2) I removed the rv_zvk file that was meant to define all Zvk<xyz> instructions. The vector crypto spec does not define groups of extensions that are declared by some overarching extension. Note that this is different from the scalar crypto specification that does define such groups. It will be simple to add such groups later if desired.

Please take another look.

I am concerned about the usability of the pseudos between P and Zvk instructions. They do allow building riscv-opcodes without an issue, however using the generated encodings in riscv-isa-sim (Spike) leads to build failures:
```
➜ make -j 1
g++ -MMD -MP  -DPREFIX=\"/scratch/install/riscv-isa-sim\" -Wall -Wno-unused -Wno-nonportable-include-path -g -O2 -fPIC -g -O2 -DPREFIX=\"/scratch/install/riscv-isa-sim\" -Wall -Wno-unused -Wno-nonportable-include-path -g -O2 -fPIC -std=c++17 -g -O2   -I. -I.. -I../fesvr -I../riscv -I../disasm -I../customext -I../fdt -I../softfloat -I../spike_main -I../spike_dasm -I/include/boost-0   -include-pch  insn_template.h.gch -fPIC -c ../riscv/processor.cc
In file included from ../riscv/processor.cc:1071:
./insn_list.h:848:1: error: use of undeclared identifier 'rstsa16_supported'
DEFINE_INSN(rstsa16)
^
../riscv/processor.cc:1058:9: note: expanded from macro 'DEFINE_INSN'
    if (name##_supported) { \
        ^
<scratch space>:67:1: note: expanded from here
rstsa16_supported
^
In file included from ../riscv/processor.cc:1071:
./insn_list.h:848:1: error: use of undeclared identifier 'rstsa16_match'
../riscv/processor.cc:1060:9: note: expanded from macro 'DEFINE_INSN'
        name##_match, \
        ^
<scratch space>:68:1: note: expanded from here
rstsa16_match
...
```

Do you have a suggestion here? Are we condemned to specifying EXTENSIONS when building?

**Comment by neelgala** [comment]
we expect both P and Zvk envodings to exist together in Spike ? If that is the case, you will need to add the list of pseudo_ops you want in encoding.out.h [here](https://github.com/riscv/riscv-opcodes/blob/master/parse.py#L959). If the list is too big you can choose to maintain it in constants.py as an array.

**Comment by egouriou-rivos** [comment]
Neel, thank you very much for the suggestion to use the `include_pseudo_ops=` argument. It worked great. With that change Spike can compile with either the extension groups "riscv_insn_ext_p" or "riscv_insn_ext_zvk". It is not ideal as it should be possible to compile both groups in and select at run time, but it might be a good starting point.

Please take another look at the new version.

**Comment by timhsu404** [comment]
@egouriou-rivos @neelgala As I know, umax32 will be conflicted with vaes*.vv instructions, such as `vaesdf.vv` in following code section shows. We need to handle these case, or it will cause some issue if we apply the opcode to spike without aware of the confliction case.
```
umax32      31..25=0b1010001   rs2  rs1        14..12=0b010   rd  6..0=0b1110111
vaesdf.vv   31..26=0x28 25=1   vs2  19..15=0x1 14..12=0x2     vd  6..0=0x77
```

**Comment by neelgala** [comment]
@egouriou-rivos can you address the case raised by @timhsu404  in this PR ?

**Comment by drom** [comment]
Does `zvkg` have 2 instructions as I see here? https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkg.adoc

**Comment by drom** [comment]
I don't two suits:
`zvkn` suite https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkn.adoc
`zvks` suite https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvks.adoc

**Comment by drom** [comment]
I think `zvkns` got renamed to `zvkned` https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkned.adoc

**Comment by egouriou** [comment]
> Does `zvkg` have 2 instructions as I see here? https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkg.adoc

Indeed. Those are present in the version I just pushed.

**Comment by egouriou** [comment]
> I think `zvkns` got renamed to `zvkned` https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkned.adoc

Indeed. Those are present in the version I just pushed.

**Comment by egouriou** [comment]
> I don't two suits:
> `zvkn` suite https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkn.adoc
> `zvks` suite https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvks.adoc

Indeed. Those are present in the version I just pushed.

**Comment by egouriou** [comment]
> @egouriou-rivos @neelgala As I know, umax32 will be conflicted with vaes*.vv instructions, such as `vaesdf.vv` in following code section shows. We need to handle these case, or it will cause some issue if we apply the opcode to spike without aware of the confliction case.
> 
> ```
> umax32      31..25=0b1010001   rs2  rs1        14..12=0b010   rd  6..0=0b1110111
> vaesdf.vv   31..26=0x28 25=1   vs2  19..15=0x1 14..12=0x2     vd  6..0=0x77
> ```

Thanks for pointing this conflict. I marked umax32 as a pseudo-op, as done with other P instructions, and added it in the list of pseudo instructions to be emitted.

**Comment by egouriou** [comment]
@neelgala, please take another look. The new version matches the current Zvk draft (Version v0.3.0, 06 February, 2023, https://github.com/riscv/riscv-crypto/releases/tag/v20230206).

@drom This version should address the issues you raised. Thanks for the feedback. Let me know if you see anything amiss.

**Comment by nibrunieAtSi5** [comment]
> > I think `zvkns` got renamed to `zvkned` https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkned.adoc
> 
> Indeed. Those are present in the version I just pushed.

Would it be worth it to also update the PR description ? (it still mentions `Zvkns`)

File: unratified/rv_zvkn
Comment (nibrunieAtSi5) [inline unratified/rv_zvkn:34]: `Zvkns` contains `Zvknhb` rather than `Zvknha` (same mnemonic as you know but extended SEW range). Since instructions were defined in `rv_zvknha` it may not be possible to export them from `rv_zvknhb` but maybe mentioning it in the comment could help.

Comment (egouriou-rivos) [reply unratified/rv_zvkn]: Done, thanks for the suggestion.

**Comment by egouriou-rivos** [comment]
> > > I think `zvkns` got renamed to `zvkned` https://github.com/riscv/riscv-crypto/blob/master/doc/vector/riscv-crypto-vector-zvkned.adoc
> > 
> > 
> > Indeed. Those are present in the version I just pushed.
> 
> Would it be worth it to also update the PR description ? (it still mentions `Zvkns`)

Done. Please take another look.

**Comment by drom** [comment]
LGTM

**Comment by egouriou-rivos** [comment]
@neelgala, @nibrunieAtSi5, I'd appreciate a review.
It is not clear to me what the merge procedure is. Feel free to trigger the merge when things look good.

File: unratified/rv_zvkn
Comment (nibrunieAtSi5) [inline unratified/rv_zvkn:32]:
```suggestion
# "Zvkn" implies "Zvknhb". We import the instructions from 'rv_zvknha',
```

**Review by nibrunieAtSi5 — APPROVED** [review]
LGTM, minor typo in a comment.

I feel Ken Dockser should be involved in this review (he is the spec editor)

File: unratified/rv_zvkn
Comment (neelgala) [reply unratified/rv_zvkn:32]: @egouriou-rivos if you fix this one comment .. then I can merge this.

Comment (egouriou-rivos) [reply unratified/rv_zvkn]: Done, thanks for catching this.

**Comment by egouriou-rivos** [comment]
I have been working closely with Ken (@kdockser) to track specification changes. He has not reviewed the patch.

I pushed a new version that updates to the latest version of the specification, as of 2023-03-22.
Major changes:
- changed Zvkb to Zvbb
- Added new instructions to Zvbb: vwsll, vbrev, vclz, vctz, vcpop
- Moved carryless multiply to it’s own extension: Zvbc

This version has a 5-bits immediate for vwsll.vi (was a 6-bits immediate for a short while).

At this point we do not expect major changes to the specification. I feel that small changes can be done on top of this commit.

**Comment by egouriou-rivos** [comment]
I am not sure how to clear the "Changes requested". As far as I know, all requested changes have been performed.

@nibrunieAtSi5, @neelgala  What are the next steps here?

**Comment by nibrunieAtSi5** [comment]
> I am not sure how to clear the "Changes requested". As far as I know, all requested changes have been performed.
> 
> @nibrunieAtSi5, @neelgala What are the next steps here?

@neelgala needs to re-review the PR and approve/request new changes.

**Comment by egouriou-rivos** [comment]
@neelgala Neel, please take another look.

File: unratified/rv_zvbb
Comment (nibrunieAtSi5) [inline unratified/rv_zvbb:34]: I think the instruction spec title was recently modified to match the mnemonic:
```suggestion
# Vector Widening Shift Left Logical
```

**Review by nibrunieAtSi5 — APPROVED** [review]
still LGTM

File: unratified/rv_zvbb
Comment (egouriou-rivos) [reply unratified/rv_zvbb]: Done, thanks for spotting this change.

**Comment by egouriou-rivos** [comment]
I pushed a new version picking the latest changes in master/, and addressing Nicolas's comment regarding the vwsll comment.

The description was updated to indicate that this is current as of version 0.9.1, 2023-04-25, Freeze Candidate.

@neelgala Please review/approve?

File: unratified/rv_zvknh
Comment (egouriou-rivos) [reply unratified/rv_zvknh]: For some reason Github keeps showing this as a needed change, so I unresolved and will mark revolved again. Let's see if this does the trick.

**Review by neelgala — APPROVED** [review]

File: unratified/rv_zvknh
Comment (nibrunieAtSi5) [reply unratified/rv_zvknh]: Yes, even if you request review again, until someone who has requested change(s) does not approve the PR it will appear as needing change.

**Pull request merged**
